cli/command/container/completion_test.go

```diff
@@ -21,6 +25,48 @@ func TestCompleteLinuxCapabilityNames(t *testing.T) {
 	}
 }
 
+func TestCompletePid(t *testing.T) {
+	tests := []struct {
+		containerListFunc   func(container.ListOptions) ([]container.Summary, error)
+		toComplete          string
+		expectedCompletions []string
+		expectedDirective   cobra.ShellCompDirective
+	}{
+		{
+			toComplete:          "",
+			expectedCompletions: []string{"container:", "host"},
+			expectedDirective:   cobra.ShellCompDirectiveNoFileComp,
+		},
+		{
+			toComplete:          "c",
+			expectedCompletions: []string{"container:"},
+			expectedDirective:   cobra.ShellCompDirectiveNoSpace,
+		},
+		{
+			containerListFunc: func(container.ListOptions) ([]container.Summary, error) {
+				return []container.Summary{
+					*builders.Container("c1"),
+					*builders.Container("c2"),
+				}, nil
+			},
+			toComplete:          "container:",
+			expectedCompletions: []string{"container:c1", "container:c2"},
+			expectedDirective:   cobra.ShellCompDirectiveNoFileComp,
+		},
+	}
+
+	for _, tc := range tests {
+		cli := test.NewFakeCli(&fakeClient{
+			containerListFunc: tc.containerListFunc,
+		})
+
+		completions, directive := completePid(cli)(NewRunCommand(cli), nil, tc.toComplete)
+
+		assert.DeepEqual(t, completions, tc.expectedCompletions)
+		assert.Equal(t, directive, tc.expectedDirective, fmt.Sprintf("wrong directive in completion for '%s'", tc.toComplete))
+	}
```

Comment:
This can use a sub-test so that we don't fail early, and each of the test-cases can be run separately. Using a sub-test also means that it's not needed to put the custom comments (`fmt.Sprintf`)  in the assertions to find "which of the tests failed";

```go
t.Run(tc.toComplete, func(t *testing.T) {
	cli := test.NewFakeCli(&fakeClient{
		containerListFunc: tc.containerListFunc,
	})

	completions, directive := completePIDMode(cli)(NewRunCommand(cli), nil, tc.toComplete)
	assert.DeepEqual(t, completions, tc.expectedCompletions)
	assert.Equal(t, directive, tc.expectedDirective)
})
```

I also tend to use `assert.Check` where possible; for cases where both results _can_ be asserted, that would allow both to be done, even if one fails (it gives slightly more information on failing tests);

```go
assert.Check(t, is.DeepEqual(completions, tc.expectedCompletions))
assert.Check(t, is.Equal(directive, tc.expectedDirective))
```